apps/extension/src/state/seed-phrase/generate.ts

```diff
@@ -4,8 +4,17 @@ import {
   generateSeedPhrase,
   SeedPhraseLength,
   ValidationField,
-  validationFields,
-} from '@penumbra-zone/crypto-web';
+} from '@penumbra-zone/crypto-web/src/mnemonic';
+
+export const generateValidationFields = (
+  seedPhrase: string[],
+  amount: number,
+): ValidationField[] => {
+  const allWords: ValidationField[] = seedPhrase.map((word, index) => ({ word, index }));
+  const shuffleWords = allWords.sort(() => 0.5 - Math.random());
+  const pickWords = shuffleWords.slice(0, amount);
+  return pickWords.sort((a, b) => a.index - b.index);
+};
```

Comment:
Why do we need to refactor the old code? Also, we seem to have lost the tests for this. Being co-located with the mnemonic stuff in the crypto package feels like a better spot.

Reply:
it was the only use of the `lodash` package in the entire monorepo, for a single random subset function. the entire lodash package was being bundled into every script on both the frontend and backend, so it was a good target to reduce js bundle size.

'validationFields' is poorly named, but its purpose is to confirm that the user has backed up their mnemonic phrase during onboarding, by satisfying a form validation prop.

it makes sense to me that we should put form validation props with the forms that they are used for. going for minimal changes, i just dropped the refactored function directly at the location of its former import. but, it should probably either stay here in the slice and be explained, or move to confirm-backup.tsx

https://github.com/penumbra-zone/web/blob/24be9588a0cf4ee5b1a29447acfed80bf9507bab/apps/extension/src/routes/page/onboarding/confirm-backup.tsx#L18-L49

the corresponding tests in the crypto package were not particularly useful, and the slice's comprehensive behavior including the functionality of this code is covered by existing tests in generate.test.ts immediately adjacent.

Reply:
> the entire lodash package was being bundled

I'm not exactly married to lodash, but this doesn't sound right to me. Webpack does tree shaking. We should only be importing the specific code for that function. Everything else should not be being pulled in.

> into every script on both the frontend and backend

By frontend, do you mean web app and backend, is the extension code? The web app does not use this code, only the extension. And given the whole bundle is pre-loaded on disk when you download it up front, it has no noticeable affect on the user's performance.

> it makes sense to me that we should put form validation props with the forms that they are used for

In theory, this would also apply to the whole of the crypto package as the password slice is the only one using most of that. The crypto package feels like a good logic unit and domain specific grouping. And if so, mnemonic validation feels properly placed in `mnemonic.ts`

Reply:
>I'm not exactly married to lodash, but this doesn't sound right to me. Webpack does tree shaking. We should only be importing the specific code for that function. Everything else should not be being pulled in. [...] By frontend, do you mean web app and backend, is the extension code? The web app does not use this code, only the extension.

haha yeah :( our current bundle config is bad. getting rid of lodash in crypto changes the bundle size of the frontend

>And given the whole bundle is pre-loaded on disk when you download it up front, it has no noticeable affect on the user's performance.

when it's in the webapp, every node will serve it.

in the extension, it was getting bundled into every script using the crypto package, which used to be most things. improving extension bundling is important because we do things like launch popups and workers that should be as fast as possible.

> In theory, this would also apply to the whole of the crypto package as the password slice is the only one using most of that.

i like this line of thought, but i'm not sure that's entirely true. let's not get ahead of ourselves

Reply:
> getting rid of lodash in crypto changes the bundle size of the frontend

Here's the reason: https://github.com/penumbra-zone/web/blob/b3a1b463bb5aec45559f4189079ac646e7c88143/apps/webapp/src/state/helpers.ts#L5

Looks like the web app is making use of one of the functions in that package. But... there's still a tree shaking issue. Nothing in that function requires lodash. I suspect it has to do with our "barrel exports" https://github.com/penumbra-zone/web/blob/main/packages/crypto/index.ts. I'm fine with getting rid of lodash, but we should investigate this when we have more time.

> improving extension bundling is important because we do things like launch popups and workers that should be as fast as possible.

For the extension, this is a function of how fast javascript can run on the machine, not network loading latency (which is usually the performance consideration of increasing the size of a bundle). The difference in the extension case is not noticeable. In fact, the most noticeable difference would be in the algorithm difference (if for some reason one was faster than the other). In any case, we are discussing the difference in single digit milliseconds.